Compare function

johankristiansson opened the pull request:
This is an alternative to #30 
The changes allow for the user to implement a custom compare function. This means that specific logic can be handled outside of the API, keeping it fresh2death.

jonaskello commented:
I don't like `allowNullOrUndefined` but since it was there from the beginning I guess we need to have it for now an deprecate it later.

@johankristiansson Could you add some tests too and update the changelog?

johankristiansson commented:
@jonaskello I agree regarding `allowNullOrUndefined` but I think we should open a different issue for that.
However, I've added tests and updated the changelog.

jonaskello commented on src/amount.ts:
```diff
@@ -13,6 +12,18 @@ export interface Amount<T extends Quantity> {
+export type Comparer = <T1 extends Quantity, T2 extends T1>(
+  left: Amount<T1>,
+  right: Amount<T2>,
+  allowNullOrUndefined: boolean
```
Perhaps `allowNullOrUndefined` could be made optional like `allowNullOrUndefined: boolean = false`. Then the external implementors of the Comparer function do not have to care/support `allowNullOrUndefined`?

johankristiansson replied:
I guess we could type it like so to make things more clear.
But in practive this is already the reality. You could define your comparer without `allowNullOrUndefined` like this:  
```
const comparer: Amount.Comparer = (
      left: Amount.Amount<Quantity.Quantity>,
      right: Amount.Amount<Quantity.Quantity>
    ) => 0;
 ```

jonaskello replied:
In that case `allowNullOrUndefined` would be `undefined` I guess as it does not have any default? If that is true the code checking `allowNullOrUndefined` would need to handle `undefined` too (in addition to true/false)?

jonaskello replied:
On second thought, `allowNullOrUndefined` is sent to each Comparer implementation so it would be up to each implementor to handle it or not. It would be nice if we somehow could make external implementors unaware of that `allowNullOrUndefined` exists at all because that would make it easier to remove it later on.

jonaskello commented:
I inspected the original code and found that `allowNullOrUndefined` is only set to true in `equals` and `compareTo`. The exported API does not expose `allowNullOrUndefined`.

Perhaps we can build the `allowNullOrUndefined` handling into the `equals` and `compareTo` functions instead of exposing it?

This is the original code:

```ts
export function equals<T1 extends Quantity, T2 extends T1>(
  left: Amount<T1>,
  right: Amount<T2>
): boolean {
  return _comparison(left, right, true) === 0;
}

export function compareTo<T1 extends Quantity, T2 extends T1>(
  left: Amount<T1>,
  right: Amount<T2>
): number {
  return _comparison(left, right, true);
}
```

The new code would be something like:

```ts
export function equals<T1 extends Quantity, T2 extends T1>(
  left: Amount<T1>,
  right: Amount<T2>
): boolean {
  // Add whatever logic allowNullOrUndefined=true had here
  if(left === undefined && right === undefined) return true;

  return _comparison(left, right) === 0;
}

export function compareTo<T1 extends Quantity, T2 extends T1>(
  left: Amount<T1>,
  right: Amount<T2>
): number {
  // Add whatever logic allowNullOrUndefined=true had here
  if(left === undefined && right === undefined) return true;

  return _comparison(left, right);
}
```

jonaskello commented:
I think the above suggestion would mean that the `allowNullOrUndefined` param could be removed from the compare function as it is now handled one level up in `equals` and `compareTo`. The other calls had `allowNullOrUndefined=false` so for them no handling is required.

johankristiansson commented:
You are correct - `allowNullOrUndefined` only exists in a mid tier in the API. No need to expose it to the outside world. I've updated the PR to match this.

Also added `comparer` to the functions `min`, `max` and `clamp` since those use compare functions internally.

jonaskello commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
+
+### Changed
+
+- All `Amount` compare functions now accepts a custom compare function as parameter.
```
I think this should go under the unrelased heading. Then later when we release we move it to a version heading.

jonaskello commented:
Looks good now :-).

jonaskello commented:
I'll merge it and you can go ahead and release it to npm.

jonaskello commented:
Or perhaps it is better you merge when you are ready :-).
